src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java

```diff
@@ -263,94 +264,150 @@ public ExamScoresDTO calculateExamScores(Long examId) {
-                                if (firstManualResult != null) {
-                                    Double resultScore = firstManualResult.getScore();
-                                    achievedPointsInFirstCorrection = resultScore != null
-                                            ? roundScoreSpecifiedByCourseSettings(resultScore / 100.0 * exercise.getMaxPoints(), exam.getCourse())
-                                            : 0.0;
-                                }
-                                studentResult.overallPointsAchievedInFirstCorrection += achievedPointsInFirstCorrection;
+        return studentExamWithGradeDTO;
+    }
+
+    @NotNull
+    private ExamScoresDTO.StudentResult calculateStudentResultWithGrade(StudentExam studentExam, List<StudentParticipation> participationsOfStudent, Exam exam,
+            ExamScoresDTO scores, ObjectMapper objectMapper) {
+        User user = studentExam.getUser();
+        var studentResult = new ExamScoresDTO.StudentResult(user.getId(), user.getName(), user.getEmail(), user.getLogin(), user.getRegistrationNumber(),
+                studentExam.isSubmitted());
+
+        studentResult.overallPointsAchieved = 0.0;
+        studentResult.overallPointsAchievedInFirstCorrection = 0.0;
+        for (StudentParticipation studentParticipation : participationsOfStudent) {
+            Exercise exercise = studentParticipation.getExercise();
+
+            // Relevant Result is already calculated
+            if (studentParticipation.getResults() != null && !studentParticipation.getResults().isEmpty()) {
+                Result relevantResult = studentParticipation.getResults().iterator().next();
+                // Note: It is important that we round on the individual exercise level first and then sum up.
+                // This is necessary so that the student arrives at the same overall result when doing their own recalculation.
+                // Let's assume that the student achieved 1.05 points in each of 5 exercises.
+                // In the client, these are now displayed rounded as 1.1 points.
+                // If the student adds up the displayed points, they get a total of 5.5 points.
+                // In order to get the same total result as the student, we have to round before summing.
+                double achievedPoints = calculateAchievedPoints(exercise, relevantResult, exam.getCourse());
+
+                // points earned in NOT_INCLUDED exercises do not count towards the students result in the exam
+                if (!exercise.getIncludedInOverallScore().equals(IncludedInOverallScore.NOT_INCLUDED)) {
+                    studentResult.overallPointsAchieved += achievedPoints;
+                }
+
+                // collect points of first correction, if a second correction exists
+                if (exam.getNumberOfCorrectionRoundsInExam() == 2 && !exercise.getIncludedInOverallScore().equals(IncludedInOverallScore.NOT_INCLUDED)) {
```

Comment:
I tried to add tests for it but it seems like this calculation requires extra information that is filtered out on before calling that method. I examined the usage and I checked the client using the related endpoints  `/courses/{courseId}/exams/{examId}/student-exams/summary` and `/courses/{courseId}/exams/{examId}/student-exams/{studentExamId}` and it seems like they do not need the first correction information so I added a flag to disable the first correction calculation when it is redundant.

Reply:
Thanks, but the flag just moves the problem. The tests don’t have to be integration tests calling the REST endpoint. Could you construct an exam that fulfils the necessary conditions so that the new method is called and then invoke `ExamService.calculateExamScores(Long examId)` directly without the REST endpoint indirection in the test?

I’m so insistent on the coverage here, because if you can’t construct such an exam, then this indicates some problem with either the covered code or the exam model:
- either some of the conditions leading up to the point where the method is actually called are wrong which leads to it never being called even if it should be, or
- they are correct and this code can indeed never be reached (maybe because the data model changed at some point) and it can be removed.

Reply:
I see, you are right. I added the necessary conditions for it to `ExamIntegrationTest.testGetExamScore`. Now it is covered and Pitest shows all lines inside the method green except for the changed conditional boundary mutation for `if (submission.getManualResults().size() > 1)` line but the lines inside the if block successfully killed mutations.